src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,725 @@
+    umad_len = IB_VENDOR_RANGE2_DATA_OFFS + data_size;
+
+    ret = umad_send(fd, agent, umad, umad_len, 0, 0);
+    if (ret < 0) {
+        ucs_error("umad_send(agent %d, umad_len %d) failed: %m", agent,
+                  umad_len);
+        status = UCS_ERR_IO_ERROR;
+        goto out;
+    }
+
+    status = UCS_OK;
+out:
+    free(umad);
+    return status;
+}
+
+static ucs_status_t perftest_mad_send(perftest_mad_rte_group_t *rte_group,
+                                      void *buffer, size_t size)
+{
+    const struct iovec iovec = {
+        .iov_base = buffer,
+        .iov_len  = size,
+    };
+    return perftest_mad_sendv(rte_group, &iovec, 1);
+}
+
+static void *realloc_or_free(void *old, size_t size)
+{
+    void *ptr = realloc(old, size);
+    if (ptr == NULL) {
+        /* cppcheck-suppress doubleFree */
+        free(old);
+    }
+    return ptr;
+}
+
+static const char ping_str[] = "ping";
+
+static ucs_status_t perftest_mad_ping(perftest_mad_rte_group_t *rte_group)
+{
+    const struct iovec iovec = {
+        .iov_base = (void*)ping_str,
+        .iov_len  = sizeof(ping_str),
+    };
+    return perftest_mad_sendv(rte_group, &iovec, 1);
+}
+
+static int perftest_mad_is_ping(void *buffer, size_t size)
+{
+    size_t ping_size = sizeof(ping_str);
+    if (size != ping_size) {
+        return 0;
+    }
+    return memcmp(buffer, ping_str, ping_size) == 0;
+}
+
+static ucs_status_t perftest_mad_recv(perftest_mad_rte_group_t *rte_group,
+                                      void *buffer, size_t *avail,
+                                      ib_portid_t *remote_port)
+{
+    int fd         = mad_rpc_portid(rte_group->mad_port);
+    size_t ref_len = *avail + IB_VENDOR_RANGE2_DATA_OFFS;
+    void *umad     = NULL;
+    int ret;
+    uint8_t *data;
+    int len; /* cannot use 'size_t' here */
+    struct ib_user_mad *user_mad;
+    ucs_status_t status;
+
+    for (;;) {
+        if ((ref_len > (INT_MAX - umad_size())) || (ref_len == 0)) {
+            ucs_error("MAD: receive: invalid umad_recv length %zu", ref_len);
+            status = UCS_ERR_INVALID_PARAM;
+            goto terminate;
+        }
+
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (umad == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+    ping_retry:
+        user_mad = umad;
+        len      = ref_len;
+
+        ret = umad_recv(fd, umad, &len, 3 * 1000);
+        if (ret >= 0) {
+            break;
+        }
+
+        if ((errno == ETIMEDOUT) || (errno == EAGAIN)) {
+            status = UCS_ERR_NO_PROGRESS;
+            goto terminate;
+        } else if ((errno != EINVAL) && (errno != ENOSPC)) {
+            ucs_error("MAD: umad_recv(fd %d, len %d) ret:%d failed: %m", fd,
+                      len, ret);
+            status = UCS_ERR_IO_ERROR;
+            goto terminate;
+        } else if (errno == EINVAL) {
+            /* EINVAL: no length info */
+            ref_len *= 2; /* 'len' has become invalid */
```

Comment:
why need to multiple length by 2?

Reply:
if I recall the case, it is when read() inside umad_recv() returns EINVAL and length is updated by umad_recv, so it cannot be reused, still needs to be grown